### PR TITLE
MINOR: enforce setting listeners in CREATE state.

### PR DESCRIPTION
**Pull request by bbejeck**

### CODE REVIEW

**Comment by bbejeck** [comment]
@dguy @guozhangwang @mjsax follow on PR from #3325 to enforce setting `UncaughtExceptionHandler` and `StateListener` only in `CREATE` state.

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 1]
```diff
@@ -737,12 +745,18 @@ public void cleanUp() {
...
+                    thread.setUncaughtExceptionHandler(eh);
+                }
 
-        if (globalStreamThread != null) {
-            globalStreamThread.setUncaughtExceptionHandler(eh);
+                if (globalStreamThread != null) {
+                    globalStreamThread.setUncaughtExceptionHandler(eh);
+                }
+            } else {
+                throw new IllegalStateException("Can only set UncaughtExceptionHandler in CREATE state.");
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: nit `CREATE` -> `CREATED`.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6322/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6306/
Test PASSed (JDK 8 and Scala 2.12).

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java (hunk 1 above)
Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Ack

**Comment by bbejeck** [comment]
@guozhangwang updated per comment.

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java [hunk 2]
```diff
@@ -254,6 +254,32 @@ public void testCannotStartTwice() throws Exception {
         }
     }
 
+    @Test(expected = IllegalStateException.class)
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java]: We should not expect an exception here but add a `fail(...)` after `streams.setUncaughtExceptionHandler(null);` -- and thus, not rethrow within catch.

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java [hunk 3]
```diff
@@ -254,6 +254,32 @@ public void testCannotStartTwice() throws Exception {
...
+            streams.setUncaughtExceptionHandler(null);
+        } catch (final IllegalStateException e) {
+            Assert.assertEquals("Can only set UncaughtExceptionHandler in CREATED state.", e.getMessage());
+            throw e;
+        } finally {
+            streams.close();
+        }
+    }
+
+    @Test(expected = IllegalStateException.class)
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java]: As above.

**Review by mjsax — CHANGES_REQUESTED** [review]
Some comments about the tests. Otherwise LGTM.

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java (hunk 2 above)
Comment (bbejeck) [reply streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java]: Ack

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java (hunk 3 above)
Comment (bbejeck) [reply streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java]: Ack

**Comment by bbejeck** [comment]
comments addressed

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6329/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6314/
Test PASSed (JDK 8 and Scala 2.12).

**Review by dguy — APPROVED** [review]
Thanks @bbejeck, LGTM

**Comment by dguy** [comment]
Merged to trunk

**Pull request closed without merging**